TUTORIAL.adoc

```diff
@@ -0,0 +1,537 @@
+
+If you just want to try Meander out, hide this tutorial with the `^` button.
+For a guided tour, press the `&gt;` button.
+
+
+== Diving in the deep end
+
+Meander contains a bunch of neat stuff.
+So much so that it may appear confusing as to where to start.
+The link:README.md[readme] starts with the fundamental building blocks and their operation.
+In contrast, this is a top down tutorial.
+We will start at the apex of Meander's power; term rewriting with the `rewrite` macro.
+
+Instead of describing an end to end translation, we will be creating small steps.
+We will write patterns and control how these patterns are applied and reapplied using combinators.
+We will cover the operators available to us inside `rewrite` expressions.
+Finally we shall look at some common situations where we can employ meander to express structural solutions,
+where otherwise we might have used linguistic approaches.
+
+The `rewrite` construct exists in the `strategy` namespace.
+The `strategy` namespace also contains the mechanisms for combining rewrite steps.
+For now it is the only namespace we need concern ourselves with.
+Pretty much anything we could do with the other namespaces, we can achieve with a `rewrite`.
+
+[source,clojure]
+----
+(ns meander.examples
+  (:require [meander.delta.strategy :as m]))
+----
+
+
+== Our first rewrite
+
+[source,clojure]
+----
+((m/rewrite
+  ?x
+  "foo")
+ "bar")
+;=> "foo"
+----
+
+Rewrite takes a match pattern and and a substitution pattern.
+In this case the match pattern is `?x` and the substitution pattern is `"foo"`.
+`?x` is a logic variable. In this pattern it will match with anything, as it stands alone.
+
+Meander's `rewrite` is a macro that constructs a function.
+Meander is compiling an efficient rewrite function for us to use with arbitrary inputs.
+Here we immediately invoke our rewriting function on the input `"bar"`.
+The result of calling the rewrite function on `"bar"` is `"foo"`.
+
+Using the rewrite immediately like this is a convenient way to try our rewrite expressions as we go.
+
+
+== Reusing rewrite
+
+Once we are happy with a transformation, it makes sense to give it a name.
+
+[source,clojure]
+----
+(def foo-bar
+  (m/rewrite
+    ?x
+    "foo"))
+
+(foo-bar "bar")
+;=> "foo"
+(foo-bar "baz")
+;=> "baz"
+----
+
+Meander compiles a function that takes some input, and rewrites it to a new output.
+
+
+== Structural pattern matching
+
+Unlike Clojure's pervasive sequence abstraction that applies to all collections,
+Meander matches structure explicitly.
+
+[source,clojure]
+----
+(def g
+  (m/rewrite
+    [?a ?b ?c ?d]
+    {?a ?c ?b ?d}))
+
+(g [1 2 3 4])
+;=> {1 3, 2 4}
+(g '(1 2 3 4))
+;=> *fail*
+----
+
+If we look for a vector, we will only match with a vector.
+When there is a match, the output is explicitly of the type that the pattern specifies.
+
+
+== Example
+
+[source,clojure]
```

Comment:
Is there a mistake with this example? I ran it and it returned `true` instead of `1`.
From what I understand, `not-not` is a regular function and its arguments are evaluated first before it is applied - quoting the argument didn't work either.

Reply:
This looks like a small mistake. It should be

```clj
(not-not 1)
;; => 1
```